test/jdk/sun/net/www/http/KeepAliveCache/B8291637.java

```diff
@@ -0,0 +1,130 @@
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8291637
+ * @run main/othervm -Dhttp.keepAlive.time.server=20 -esa -ea B8291637
+ */
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.*;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.CompletableFuture;
+
+public class B8291637 {
+    static CompletableFuture<Boolean> passed = new CompletableFuture<>();
+
+    static class Server extends Thread {
+        final ServerSocket serverSocket;
+        final int port;
+        volatile Socket s;
+
+        public Server() throws IOException {
+            serverSocket = new ServerSocket(0);
+            port = serverSocket.getLocalPort();
+            setDaemon(true);
+        }
+
+        public int getPort() {
+            return port;
+        }
+
+        public void close() {
+            try {
+                serverSocket.close();
+                s.close();
+            } catch (IOException e) {}
+        }
+
+        static void readRequest(Socket s) throws IOException {
+            InputStream is = s.getInputStream();
+            is.read();
+            while (is.available() > 0)
+                is.read();
+        }
+
+        public void run() {
+            try {
+                while (true) {
+                    s = serverSocket.accept();
+                    readRequest(s);
+                    OutputStream os = s.getOutputStream();
+                    String resp = "" +
+                            "HTTP/1.1 200 OK\r\n" +
+                            "Content-Length: 11\r\n" +
+                            "Connection: Keep-Alive\r\n" +
+                            "Keep-Alive: timeout=-10\r\n" + // invalid negative value
+                            "\r\n" +
+                            "Hello World";
+                    os.write(resp.getBytes(StandardCharsets.ISO_8859_1));
+                    os.flush();
+                    InputStream is = s.getInputStream();
+                    long l1 = System.currentTimeMillis();
+                    is.read();
+                    long l2 = System.currentTimeMillis();
+                    long diff = (l2 - l1) / 1000;
+                    /*
+                     * timeout is set to 20 seconds. If bug is still present
+                     * then the timeout will occur the first time the keep alive
+                     * thread wakes up which is after 5 seconds. This allows
+                     * very large leeway with slow running hardware.
+                     */
+                    if (diff < 19) {
+                        passed.complete(false);
+                    } else {
+                        passed.complete(true);
+                    }
+                    System.out.println("Time diff = " + diff);
+                }
+            } catch (IOException e) {
+                System.err.println("Server exception terminating");
```

Comment:
It looks to me that if an exception occurs (not just `IOException`) then this test might hang (and timeout with the jtreg timeout) at the `passed.get()` call in the `main()` method. Should we perhaps catch `Throwable` here and do `passed.completeExceptionally`?

Reply:
Seems reasonable.